codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/RestJsonProtocolGenerator.java

```diff
@@ -374,4 +374,762 @@ private String getSerializedMemberName(MemberShape memberShape) {
+    }
+
+
+    // Generates deserializer for String member shape.
+    private String generateDocumentBindingStringMemberDeserializer(
+            GoWriter writer,
+            Model model,
+            SymbolProvider symbolProvider,
+            MemberShape memberShape
+    ) {
+        String memberName = symbolProvider.toMemberName(memberShape);
+        Shape targetShape = model.expectShape(memberShape.getTarget());
+        Symbol targetSymbol = symbolProvider.toSymbol(targetShape);
+        writer.write("val, err := decoder.Token()");
+        writer.write("if err != nil { return err }");
+        writer.write("st, ok := val.(string)");
+        writer.openBlock("if !ok {", "}", () -> {
+            writer.write("return fmt.Errorf(\"expected $L to be of type $P, got %T instead\", st)"
+                    , memberName, targetSymbol);
+        });
+
+        if (targetShape.hasTrait(EnumTrait.class)) {
+            return String.format("types.%s(st)", targetSymbol.getName());
+        }
+
+        return CodegenUtils.generatePointerReferenceIfPointable(memberShape, "st");
+    }
+
+    // Generates deserializer for Boolean member shape.
+    private String generateDocumentBindingBooleanMemberDeserializer(
+            GoWriter writer,
+            SymbolProvider symbolProvider,
+            MemberShape memberShape
+    ) {
+        String shapeName = symbolProvider.toMemberName(memberShape);
+        Symbol shapeSymbol = symbolProvider.toSymbol(memberShape);
+        writer.write("val, err := decoder.Token()");
+        writer.write("if err != nil { return err }");
+        writer.write("b, ok := val.(bool)");
+        writer.openBlock("if !ok {", "}", () -> {
+            writer.write("return fmt.Errorf(\"expected $L to be of type $L, got %T instead\", st)"
+                    , shapeName, shapeSymbol.getName());
+        });
+
+        return CodegenUtils.generatePointerReferenceIfPointable(memberShape, "b");
+    }
+
+    // Generates deserializer for Byte, Short, Integer, Long, Big Integer member shape.
+    private String generateDocumentBindingIntegerMemberDeserializer(
+            GoWriter writer,
+            MemberShape memberShape
+    ) {
+        writer.write("val, err := decoder.Token()");
+        writer.write("if err != nil { return err }");
+        writer.write("nt, err := val.(json.Number).Int64()");
+        writer.write("if err != nil { return err }");
+        switch (memberShape.getType()) {
+            case BYTE:
+                writer.write("st := byte(nt)");
+                break;
+            case SHORT:
+                writer.write("st := int16(nt)");
+            case INTEGER:
+                writer.write("st := int32(nt)");
+                break;
+            case LONG:
+                writer.write("st := nt");
+                break;
+            case BIG_INTEGER:
+                writer.addUseImports(GoDependency.BIG);
+                writer.write("st := big.NewInt(nt)");
+                break;
```

Comment:
It might be safer for big int to use `i, ok := new(big.Int).SetString(val.(string), 10)` instead of truncating to int64.

This will prevent unexpected behavior and aligns with the SDK's serialization strategy with regard to JSON and bigs.